### PR TITLE
<enhance>imeplement_mqtt_telemetry_service

### PR DESCRIPTION
**Pull request by MrLeea-13155bc**
**What this PR does / why we need it**:
imeplement_mqtt_telemetry_service
**Will this PR make the community happier**? 
yes
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**How is this PR tested**
- [x] unit test
- [x] e2e test
- [ ] other (please specify)

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: pkg/deviceshifu/deviceshifubase/push_telemetry.go [hunk 1]
```diff
@@ -20,33 +40,79 @@ func CopyHeader(dst, src http.Header) {
...
+		return err
+	}
+	request := TelemetryRequest{
+		RawData:     rawData,
+		MQTTSetting: settings.ServiceSettings.MQTTSetting,
 	}
+
+	requestBody, err := json.Marshal(request)
+	if err != nil {
+		klog.Errorf("Error when marshal request to []bute, error: %v", err)
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifubase/push_telemetry.go:80]: byte

File: pkg/deviceshifu/deviceshifubase/push_telemetry_test.go [hunk 2]
```diff
@@ -276,15 +295,17 @@ func TestGetTelemetryCollectionServiceMap(t *testing.T) {
...
 			"",
 		},
 	}
 
 	for _, c := range testCases {
 		t.Run(c.Name, func(t *testing.T) {
 			result, err := getTelemetryCollectionServiceMap(c.inputDevice)
-			assert.Equal(t, c.expectedMap, result)
+			ok := assert.ObjectsAreEqual(c.expectedMap, result)
+			log.Printf("%#v\n%#v\n%#v", c.Name, c.expectedMap, result)
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifubase/push_telemetry_test.go:307]: klog

File: pkg/deviceshifu/deviceshifubase/push_telemetry_test.go [hunk 3]
```diff
@@ -311,5 +332,125 @@ func TestPushToHTTPTelemetryCollectionService(t *testing.T) {
...
+			settings: &v1alpha1.TelemetryServiceSpec{
+				ServiceSettings: &v1alpha1.ServiceSettings{
+					MQTTSetting: &v1alpha1.MQTTSetting{
+						MQTTTopic: unitest.ToPointer("/test/topic"),
+					},
+				},
+				Address: unitest.ToPointer(address),
+			},
+			expectedErr: "",
+		},
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifubase/push_telemetry_test.go:392]: duplicated

File: pkg/deviceshifu/deviceshifubase/push_telemetry_test.go [hunk 4]
```diff
@@ -311,5 +332,125 @@ func TestPushToHTTPTelemetryCollectionService(t *testing.T) {
...
+
+func mockMQTTTelemetryServiceServer(t *testing.T) *httptest.Server {
+	handler := http.NewServeMux()
+	handler.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+	})
+	return httptest.NewServer(handler)
+}
+
+func TestPushTElemetryCollectionService(t *testing.T) {
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifubase/push_telemetry_test.go:412]: TestPushTelemetryCollectionService

File: pkg/deviceshifu/deviceshifubase/push_telemetry_test.go [hunk 5]
```diff
@@ -311,5 +332,125 @@ func TestPushToHTTPTelemetryCollectionService(t *testing.T) {
...
+					},
+				},
+				Address:  unitest.ToPointer(address),
+				Protocol: unitest.ToPointer(v1alpha1.ProtocolMQTT),
+			},
+			message: &http.Response{
+				Body: io.NopCloser(bytes.NewBufferString("test")),
+			},
+			expectedErr: "",
+		},
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifubase/push_telemetry_test.go:447]: let's add one with not supported protocol

**Review by tomqin93 — COMMENTED** [review]
@MrLeea-13155bc added some comment

File: Makefile [hunk 6]
```diff
@@ -45,6 +45,11 @@ buildx-push-image-deviceshifu: \
 	buildx-push-image-deviceshifu-http-socket \
 	buildx-push-image-deviceshifu-http-opcua
 
+buildx-push-image-telemetryServiceMQTT:
```
Comment (BtXin) [inline Makefile:48]: Remove `MQTT` at the end

File: Makefile [hunk 7]
```diff
@@ -77,6 +82,11 @@ buildx-build-image-deviceshifu: \
 	buildx-build-image-deviceshifu-http-opcua \
 	buildx-build-image-deviceshifu-http-plc4x 
 
+buildx-build-image-telemetryServiceMQTT:
```
Comment (BtXin) [inline Makefile:85]: ditto

File: Makefile [hunk 8]
```diff
@@ -45,6 +45,11 @@ buildx-push-image-deviceshifu: \
 	buildx-push-image-deviceshifu-http-socket \
 	buildx-push-image-deviceshifu-http-opcua
 
+buildx-push-image-telemetryService:
```
Comment (tomqin93) [inline Makefile:48]: let's be consistent and use `buildx-push-image-telemetry-service` instead.
Also the push statement should include all three platforms `--platform=linux/amd64,linux/arm64,linux/arm`

File: Makefile [hunk 9]
```diff
@@ -77,6 +82,11 @@ buildx-build-image-deviceshifu: \
 	buildx-build-image-deviceshifu-http-opcua \
 	buildx-build-image-deviceshifu-http-plc4x 
 
+buildx-build-image-telemetryService:
```
Comment (tomqin93) [inline Makefile:85]: same as above

File: azure-pipelines/azure-pipelines.yml [hunk 10]
```diff
@@ -444,6 +444,44 @@ stages:
...
+        inputs:
+          dockerVersion: 20.10.9
+          releaseType: stable
+      - script: |
+          docker buildx create --use
+          docker buildx build --platform=linux/amd64 \
+          -f $(System.DefaultWorkingDirectory)/dockerfiles/Dockerfile.telemetryservice \
+          --build-arg PROJECT_ROOT="$(System.DefaultWorkingDirectory)" $(System.DefaultWorkingDirectory) \
+          -t edgehub/telemetryservice:$(tag) --load
+        displayName: build edgehub/dtelemetryService
```
Comment (tomqin93) [inline azure-pipelines/azure-pipelines.yml:463]: dtelemetryService remove d

File: azure-pipelines/azure-pipelines.yml [hunk 11]
```diff
@@ -444,6 +444,44 @@ stages:
...
+          docker buildx build -f --platform=linux/amd64 \
+          -f $(System.DefaultWorkingDirectory)/examples/telemetryservice/mockserver/Dockerfile.mockserver \
+          --build-arg PROJECT_ROOT="$(System.DefaultWorkingDirectory)" $(System.DefaultWorkingDirectory) \
+          -t edgehub/mockserver:$(tag) --load
+        displayName: build mockServer
+      - script: |
+          set -e
+          docker run -itd --network host --name mosquitto eclipse-mosquitto:2.0.14
+          docker run -itd --network host edgehub/telemetryservice:nightly
+          docker run -itd --network host --name mockserver edgehub/mockserver:nightly
```
Comment (tomqin93) [inline azure-pipelines/azure-pipelines.yml:480]: substitute nightly to $(tag)

File: azure-pipelines/azure-pipelines.yml [hunk 12]
```diff
@@ -686,6 +724,24 @@ stages:
...
+        inputs:
+          command: login
+          containerRegistry: dockerhub-connection
+      - script: |
+          docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+          docker buildx create --use
+        displayName: configure multi-arch and buildx
+      - script: |
+          docker buildx build --platform=linux/amd64,linux/arm64,linux/arm \
+          -f $(System.DefaultWorkingDirectory)/dockerfiles/Dockerfile.deviceshifu \
```
Comment (tomqin93) [inline azure-pipelines/azure-pipelines.yml:741]: wrong image here

File: dockerfiles/Dockerfile.telemetryservice [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+RUN CGO_ENABLED=0 GOOS=$TARGETOS GOARCH=$TARGETARCH go build -a -o /output/telemetryservice cmd/telemetryservice/main.go
+
+FROM gcr.io/distroless/static-debian11
+WORKDIR /
+COPY --from=builder /output/telemetryservice  telemetryservice 
+
+ENV SERVER_LISTEN_PORT=:17772
+
+EXPOSE 17772
```
Comment (tomqin93) [inline dockerfiles/Dockerfile.telemetryservice:29]: remove and use on demand

File: examples/telemetryservice/checkoutput.sh [hunk 14]
```diff
@@ -0,0 +1,17 @@
+#!bin/bash
+
+default='empty'
+
+for i in {1..100} 
+do
+    docker run -itd --network host edgehub/mockclient:nightly
```
Comment (tomqin93) [inline examples/telemetryservice/checkoutput.sh:7]: use tag instead of nightly

File: go.sum [hunk 15]
```diff
@@ -249,6 +249,10 @@ github.com/grpc-ecosystem/grpc-gateway v1.16.0/go.mod h1:BDjrQk3hbvj6Nolgz8mAMFb
 github.com/hashicorp/golang-lru v0.5.0/go.mod h1:/m3WP610KZHVQ1SGc6re/UDhFvYD7pJ4Ao+sR/qLZy8=
 github.com/hashicorp/golang-lru v0.5.1/go.mod h1:/m3WP610KZHVQ1SGc6re/UDhFvYD7pJ4Ao+sR/qLZy8=
 github.com/hpcloud/tail v1.0.0/go.mod h1:ab1qPbhIpdTxEkNHXyeSf5vhxWSCs/tWer42PpOxQnU=
+github.com/huin/gobinarytest v0.0.0-20170803182140-bc6c19e91749 h1:uEtvxEWZtnFRuUUyJl214tS14+BMi7iO7up/Rx4CejY=
+github.com/huin/gobinarytest v0.0.0-20170803182140-bc6c19e91749/go.mod h1:NLS5Is63wKCuzcJLN28yt3sm/Ps9OXNd9GPUlO4wfFA=
+github.com/huin/mqtt v0.0.0-20220729180547-9e81042af5f7 h1:FbE0sTosmODWNX7oRNCd+xNK8HI1Ys2Je+612xmad38=
+github.com/huin/mqtt v0.0.0-20220729180547-9e81042af5f7/go.mod h1:zbO4h3NDGxpAVWGYzQNle+BDfuwm8fWcgVxFuIHyuas=
```
Comment (tomqin93) [inline go.sum:255]: try go tidy

File: pkg/deviceshifu/deviceshifubase/push_telemetry.go [hunk 16]
```diff
@@ -1,15 +1,35 @@
...
+}
+
+func PushTelemetryCollectionService(tss *v1alpha1.TelemetryServiceSpec, message *http.Response) error {
+	var err error
+	switch *tss.Protocol {
+	case v1alpha1.ProtocolHTTP:
+		err = pushToHTTPTelemetryCollectionService(*tss.Protocol, message, *tss.Address)
+	case v1alpha1.ProtocolMQTT:
+		err = pushToMQTTTelemetryCollectionService(message, tss)
+	}
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifubase/push_telemetry.go:28]: let's add default case to return unsupported protocol

File: pkg/deviceshifu/deviceshifubase/push_telemetry.go [hunk 17]
```diff
@@ -1,15 +1,35 @@
...
+
+	return err
+}
+
 // CopyHeader HTTP header type:
 // type Header map[string][]string
 func CopyHeader(dst, src http.Header) {
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifubase/push_telemetry.go:35]: let's delete this part

File: pkg/deviceshifu/deviceshifubase/push_telemetry.go [hunk 18]
```diff
@@ -20,33 +40,79 @@ func CopyHeader(dst, src http.Header) {
...
+		klog.Errorf("Error when build request with requestBody, error: %v", err)
+		return err
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		klog.Errorf("Error when send request to Server, error: %v", err)
+		return err
+	}
+	klog.Infof("Success to Send message to %v,message: %v", string(rawData))
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifubase/push_telemetry.go:95]: successfully sent message %v to telemetry service address %v

File: pkg/deviceshifu/deviceshifubase/push_telemetry.go [hunk 19]
```diff
@@ -1,15 +1,35 @@
 package deviceshifubase
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifubase/push_telemetry.go:1]: remove underscrol in filename

File: pkg/deviceshifu/deviceshifubase/push_telemetry_test.go [hunk 20]
```diff
@@ -1,6 +1,7 @@
 package deviceshifubase
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifubase/push_telemetry_test.go:1]: remove underscrol in filename

File: pkg/deviceshifu/deviceshifubase/push_telemetry_test.go [hunk 21]
```diff
@@ -276,15 +295,17 @@ func TestGetTelemetryCollectionServiceMap(t *testing.T) {
...
 			"",
 		},
 	}
 
 	for _, c := range testCases {
 		t.Run(c.Name, func(t *testing.T) {
 			result, err := getTelemetryCollectionServiceMap(c.inputDevice)
-			assert.Equal(t, c.expectedMap, result)
+			ok := assert.ObjectsAreEqual(c.expectedMap, result)
+			klog.Infof("%#v\n%#v\n%#v", c.Name, c.expectedMap, result)
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifubase/push_telemetry_test.go:307]: let's improve the readability for this klog

File: pkg/telemetryservice/mqtt/telemetryservicemqtt.go [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+	if err != nil {
+		klog.Errorf("Error to connect to mqtt server, error: %#v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer (*client).Disconnect(0)
+
+	token := (*client).Publish(*telemetryRequest.MQTTSetting.MQTTTopic, 1, false, telemetryRequest.RawData)
+	if token.Error() != nil {
+		klog.Errorf("Error when publish Data to MqttServer, error: %#v", err.Error())
```
Comment (tomqin93) [inline pkg/telemetryservice/mqtt/telemetryservicemqtt.go:43]: Mqtt => MQTT

**Review by tomqin93 — COMMENTED** [review]
@MrLeea-13155bc added some comment ,overall looks good

**Review by tomqin93 — APPROVED** [review]
LGTM

**Pull request merged**